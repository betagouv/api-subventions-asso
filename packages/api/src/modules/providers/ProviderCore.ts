import { ProviderEnum } from "../../@enums/ProviderEnum";
import ProviderRequestFactory, { ProviderRequestService } from "../provider-request/providerRequest.service";
import Provider from "./@types/IProvider";

export default abstract class ProviderCore implements Provider {
    public http: ProviderRequestService;

    constructor(public readonly provider: { name: string; type: ProviderEnum; description: string; id: string }) {
        this.http = ProviderRequestFactory(this.provider.id);
    }
}
