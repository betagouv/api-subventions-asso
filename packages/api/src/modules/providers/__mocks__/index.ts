import { IncomingMessage } from "http";
import { Socket } from "node:net";
import { ProviderEnum } from "../../../@enums/ProviderEnum";
import { ProviderRequestService } from "../../provider-request/providerRequest.service";
import {
    applicationProvidersFixtures,
    fullGrantProvidersFixtures,
    grantProvidersFixtures,
    paymentProvidersFixtures,
} from "../__fixtures__/providers.fixture";

const providers = {
    // RAW service AssociationProvider + DemandeSubventionProvider
    serviceA: {
        provider: {
            name: "serviceA",
            id: "prov-A",
            type: ProviderEnum.raw,
            description: "descriptionA",
        },
        isAssociationsProvider: true,
        isDemandesSubventionsProvider: true,
        getAssociationsBySiren: () => [{}],
        getAssociationsBySiret: () => [{}],
        getAssociationsByRna: () => [{}],
        getDemandeSubventionBySiret: async () => [{}],
        getDemandeSubventionBySiren: async () => [{}],
        http: new ProviderRequestService("prov-A"),
    },
    // API service AssociationProvider + DemandeSubventionProvider
    serviceB: {
        provider: {
            name: "serviceB",
            id: "prov-B",
            type: ProviderEnum.api,
            description: "descriptionB",
        },
        isAssociationsProvider: true,
        isDemandesSubventionsProvider: true,
        getAssociationsBySiren: () => [{}],
        getAssociationsBySiret: () => [{}],
        getAssociationsByRna: () => [{}],
        getDemandeSubventionBySiret: async () => [{}],
        getDemandeSubventionBySiren: async () => [{}],
    },
    // API service AssociationProvider that returns null for all getAssociations()
    serviceC: {
        provider: {
            name: "serviceC",
            id: "prov-C",
            type: ProviderEnum.api,
            description: "descriptionC",
        },
        isAssociationsProvider: true,
        isDemandesSubventionsProvider: false,
        getAssociationsBySiren: async () => null,
        getAssociationsBySiret: async () => null,
        getAssociationsByRna: async () => null,
    },
    // API service that is not either an AssociationProvider or DemandesSubventionsProvider
    serviceD: {
        provider: {
            name: "serviceD",
            id: "prov-D",
            type: ProviderEnum.api,
            description: "descriptionD",
        },
        isAssociationsProvider: false,
        isDemandesSubventionsProvider: false,
        getAssociationsBySiren: async () => [{}],
        getAssociationsBySiret: async () => [{}],
        getAssociationsByRna: async () => [{}],
    },
    // RAW sercice that is not either an AssociationProvider or DemandesSubventionsProvider
    serviceE: {
        provider: {
            name: "serviceE",
            id: "prov-E",
            type: ProviderEnum.raw,
            description: "descriptionE with getSpecificDocumentStream",
        },
        isAssociationsProvider: false,
        isDemandesSubventionsProvider: false,
        getAssociationsBySiren: async () => [{}],
        getAssociationsBySiret: async () => [{}],
        getAssociationsByRna: async () => [{}],
        getSpecificDocumentStream: async (_path: string) => new IncomingMessage(new Socket()),
    },
};

export default providers;

export const fullGrantProviders = fullGrantProvidersFixtures;
export const demandesSubventionsProviders = applicationProvidersFixtures;
export const paymentProviders = paymentProvidersFixtures;
export const grantProviders = grantProvidersFixtures;
