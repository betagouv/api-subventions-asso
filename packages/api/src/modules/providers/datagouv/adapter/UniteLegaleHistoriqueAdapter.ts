import AssociationNameEntity from "../../../association-name/entities/AssociationNameEntity";
import { UniteLegalHistoryRaw } from "../@types/UniteLegalHistoryRaw";
import dataGouvService from "../datagouv.service";
import EntrepriseSirenEntity from "../entities/EntrepriseSirenEntity";

export class UniteLegaleHistoriqueAdapter {
    static rawToAssociationName(raw: UniteLegalHistoryRaw) {
        return new AssociationNameEntity(
            null,
            raw.denominationUniteLegale,
            dataGouvService.provider.name,
            new Date(raw.dateDebut),
            raw.siren,
        );
    }

    static rawToEntrepriseSiren(raw: UniteLegalHistoryRaw) {
        return new EntrepriseSirenEntity(raw.siren);
    }
}
