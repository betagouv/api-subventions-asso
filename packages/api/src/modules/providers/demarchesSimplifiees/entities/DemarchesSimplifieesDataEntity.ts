import { Rna, Siret } from "@api-subventions-asso/dto";
import { DefaultObject } from "../../../../@types";

export default interface DemarchesSimplifieesDataEntity {
    siret: Siret;
    demarcheId: number;
    demande: {
        id: string;
        demandeur: {
            siret: Siret;
            association?: {
                rna: Rna;
                titre: string;
            };
        };
        demarche: {
            title: string;
        };
        groupeInstructeur: {
            label: string;
        };
        motivation: string | null;
        state: string | null;
        dateDepot: string | null;
        datePassageEnInstruction: string | null;
        dateDerniereModification: string;
        dateTraitement: string | null;
        pdf: {
            url: string;
            filename: string;
            contentType: string;
        };
        champs: DefaultObject<string>;
        annotations: DefaultObject<string>;
    };
}
