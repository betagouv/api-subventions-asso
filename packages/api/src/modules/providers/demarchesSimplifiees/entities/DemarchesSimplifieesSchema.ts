export default interface DemarchesSimplifieesSchema {
    demarcheId: number;
    schema: DemarchesSimplifieesSchemaLine[];
    commonSchema?: DemarchesSimplifieesSchemaLine[];
}

export type DemarchesSimplifieesSchemaLine = { to: string; from: string } | { to: string; value: string };
