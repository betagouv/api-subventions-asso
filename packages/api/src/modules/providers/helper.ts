import { ApplicationStatus } from "@api-subventions-asso/dto";

export function toStatusFactory(statusMap: { [K in ApplicationStatus]?: string[] }) {
    function toStatus(providerStatus: string): ApplicationStatus {
        for (const [normalizedStatus, providerStatusList] of Object.entries(statusMap)) {
            if (providerStatusList.includes(providerStatus)) return normalizedStatus as ApplicationStatus;
        }
        return ApplicationStatus.UNKNWON;
    }

    return toStatus;
}
