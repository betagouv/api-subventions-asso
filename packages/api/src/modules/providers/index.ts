import dataEntrepriseService from "./dataEntreprise/dataEntreprise.service"
import leCompteAssoService from "./leCompteAsso/leCompteAsso.service"
import osirisService from "./osiris/osiris.service"
import fonjepService from "./fonjep/fonjep.service"
import chorusService from "./chorus/chorus.service"
import gisproService from "./gispro/gispro.service"

export default {
    dataEntrepriseService,
    osirisService,
    leCompteAssoService,
    fonjepService,
    chorusService,
    gisproService
}