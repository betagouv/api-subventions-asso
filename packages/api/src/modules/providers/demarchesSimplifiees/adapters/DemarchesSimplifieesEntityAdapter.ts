import moment from "moment";
import * as lodash from "lodash";
import { ApplicationStatus, ProviderValue } from "dto";
import DemarchesSimplifieesDataEntity from "../entities/DemarchesSimplifieesDataEntity";
import DemarchesSimplifieesSchema from "../entities/DemarchesSimplifieesSchema";
import { isValidDate } from "../../../../shared/helpers/DateHelper";
import { stringIsFloat } from "../../../../shared/helpers/StringHelper";
import { DefaultObject } from "../../../../@types";
import { toStatusFactory } from "../../providers.adapter";
import { ApplicationFlatEntity } from "../../../../entities/ApplicationFlatEntity";
import { InternalServerError } from "core";
import Siret from "../../../../identifierObjects/Siret";

export class DemarchesSimplifieesEntityAdapter {
    private static _statusConversionArray: { label: ApplicationStatus; providerStatusList: string[] }[] = [
        { label: ApplicationStatus.REFUSED, providerStatusList: ["refuse"] },
        { label: ApplicationStatus.GRANTED, providerStatusList: ["accepte"] },
        { label: ApplicationStatus.INELIGIBLE, providerStatusList: ["sans_suite"] },
        { label: ApplicationStatus.PENDING, providerStatusList: ["en_instruction"] },
    ];
    private static baseFlatWithNull: Partial<ApplicationFlatEntity> = {
        allocatorId: null,
        allocatorIdType: null,
        allocatorName: null,
        budgetaryYear: null,
        cofinancersIdType: null,
        cofinancersNames: null,
        cofinancingRequested: null,
        confinancersId: null,
        conventionDate: null,
        decisionDate: null,
        decisionReference: null,
        depositDate: null,
        ej: null,
        grantedAmount: null,
        idRAE: null,
        instructiveDepartementId: null,
        instructiveDepartmentIdType: null,
        instructiveDepartmentName: null,
        joinKeyDesc: null,
        joinKeyId: null,
        managingAuthorityId: null,
        managingAuthorityIdType: null,
        managingAuthorityName: null,
        nature: null,
        object: null,
        paymentCondition: null,
        paymentConditionDesc: null,
        paymentId: null,
        paymentPeriodDates: null,
        pluriannual: null,
        pluriannualYears: null,
        requestYear: null,
        requestedAmount: null,
        scheme: null,
        subScheme: null,
        subventionPercentage: null,
        totalAmount: null,
        ueNotification: null,
    };

    private static mapSchema<T>(
        entity: DemarchesSimplifieesDataEntity,
        schema: DemarchesSimplifieesSchema,
        schemaId: string,
    ): T {
        const subvention = {};
        if (!schema[schemaId]) {
            throw new InternalServerError(`no schema for type ${schemaId} and form ${schema.demarcheId}`);
        }

        schema[schemaId].forEach(property => {
            if (property.value) return lodash.set(subvention, property.to, property.value);

            let value = lodash.get(entity, property.from);
            const valueDate = [
                moment(value, "DD MMMM YYYY", "fr", true).toDate(), // Use moment for read French date
                moment(value, true), // use moment with strict params true, to force the reading of a date in js format and not an interpretation
                new Date(value),
            ].find(date => isValidDate(date));

            if (value === undefined || value === "") return;
            else if (stringIsFloat(value)) value = parseFloat(value);
            else value = valueDate || value;

            lodash.set(subvention, property.to, value);
        });

        return subvention as unknown as T;
    }

    private static nestedToProviderValues(object: object, toPv: (v: unknown) => ProviderValue<unknown>) {
        if (Array.isArray(object)) return object.map(v => this.nestedToProviderValues(v, toPv));
        if (object.constructor !== Object) return toPv(object);
        const res = {};
        for (const [key, value] of Object.entries(object)) {
            res[key] = DemarchesSimplifieesEntityAdapter.nestedToProviderValues(value, toPv);
        }
        return res;
    }

    static toFlat(entity: DemarchesSimplifieesDataEntity, schema: DemarchesSimplifieesSchema): ApplicationFlatEntity {
        const application: DefaultObject = {
            ...DemarchesSimplifieesEntityAdapter.baseFlatWithNull,
            ...DemarchesSimplifieesEntityAdapter.mapSchema(entity, schema, "flatSchema"),
        };

        application.statusLabel = toStatusFactory(DemarchesSimplifieesEntityAdapter._statusConversionArray)(
            application.status as string,
        );
        delete application.status;

        // TODO should we try better to have exercise ?

        application.beneficiaryEstablishmentIdType = Siret.getName();
        application.provider = `demarches-simplifiees-${entity.demarcheId}`;
        application.beneficiaryEstablishmentId = (application.beneficiaryEstablishmentId as number).toString();
        application.applicationId = `${application.provider}-${application.applicationProviderId}`;
        application.uniqueId = `${application.applicationId}-${application.budgetaryYear}`;
        application.paymentId = `${application.beneficiaryEstablishmentId}-${application.ej}-${application.budgetaryYear}`; //siret-EJ-exerciceBudgetaire
        application.depositDate = application.depositDate ? new Date(application.depositDate as string) : null;
        application.requestYear = application.depositDate ? (application.depositDate as Date).getFullYear() : null;

        return application as unknown as ApplicationFlatEntity;
    }
}
