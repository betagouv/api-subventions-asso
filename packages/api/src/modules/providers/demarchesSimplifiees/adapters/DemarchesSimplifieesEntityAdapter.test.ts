import DemarchesSimplifieesDataEntity from "../entities/DemarchesSimplifieesDataEntity";
import { DemarchesSimplifieesEntityAdapter } from "./DemarchesSimplifieesEntityAdapter";
import {
    DATA_ENTITIES,
    SCHEMAS,
} from "../../../../../tests/dataProviders/db/__fixtures__/demarchesSimplifiees.fixtures";
import demarchesSimplifieesService from "../demarchesSimplifiees.service";
import DemarchesSimplifieesSchema from "../entities/DemarchesSimplifieesSchema";
import * as providerAdapter from "../../providers.adapter";

const mockAdaptStatus = jest.fn();

jest.mock("../demarchesSimplifiees.service");
jest.mock("../../providers.adapter", () => ({
    toStatusFactory: jest.fn(_conversionArray => mockAdaptStatus),
}));

describe("DemarchesSimplifieesEntityAdapter", () => {
    // @ts-expect-error: mock private method
    const mapMock: jest.SpyInstance = jest.spyOn(DemarchesSimplifieesEntityAdapter, "mapSchema");
    const SIRET = "00000000000000";
    const DEMANDE = {
        siret: SIRET,
        demande: { dateDerniereModification: new Date() },
        toto: ["jedusor"],
    } as unknown as DemarchesSimplifieesDataEntity;
    const MAPPING = {
        demarcheId: 12345,
        schema: [
            {
                from: "toto[0]",
                to: "tom",
            },
        ],
    };

    describe("rawToApplication", () => {
        // @ts-expect-error: parameter type
        const RAW_APPLICATION: RawApplication = { data: { entity: { foo: "bar" }, schema: { boo: "faz" } } };
        // @ts-expect-error: parameter type
        const APPLICATION: DemandeSubvention = { foo: "bar" };
        let mockToSubvention: jest.SpyInstance;

        beforeAll(() => {
            mockToSubvention = jest.spyOn(DemarchesSimplifieesEntityAdapter, "toSubvention");
            mockToSubvention.mockReturnValue(APPLICATION);
        });

        afterAll(() => {
            mockToSubvention.mockRestore();
        });

        it("should call toSubvention", () => {
            DemarchesSimplifieesEntityAdapter.rawToApplication(RAW_APPLICATION);
            const { entity, schema } = RAW_APPLICATION.data;
            expect(mockToSubvention).toHaveBeenCalledWith(entity, schema);
        });

        it("should return DemandeSubvention", () => {
            const expected = APPLICATION;
            const actual = DemarchesSimplifieesEntityAdapter.rawToApplication(RAW_APPLICATION);
            expect(actual).toEqual(expected);
        });
    });

    describe("toSubvention", () => {
        it("should return subvention with siret", () => {
            const actual = DemarchesSimplifieesEntityAdapter.toSubvention(DEMANDE, MAPPING);

            expect(actual.siret.value).toBe(SIRET);
        });

        it("sets siret from mapping if given", () => {
            const OTHER_SIRET = "09876543210987";
            mapMock.mockReturnValueOnce({ siret: OTHER_SIRET });
            const expected = OTHER_SIRET;
            const actual = DemarchesSimplifieesEntityAdapter.toSubvention(DEMANDE, MAPPING).siret.value;

            expect(actual).toBe(expected);
        });

        it("should use schema to build subvention", () => {
            const actual = DemarchesSimplifieesEntityAdapter.toSubvention(DEMANDE, MAPPING);

            expect(actual).toEqual(
                expect.objectContaining({
                    tom: expect.objectContaining({
                        value: "jedusor",
                    }),
                }),
            );
        });

        it("sets annee_demande thanks to date_debut", () => {
            mapMock.mockReturnValueOnce({ date_debut: new Date(2023, 10, 20) });
            const expected = 2023;
            const actual = DemarchesSimplifieesEntityAdapter.toSubvention(DEMANDE, MAPPING).annee_demande?.value;

            expect(actual).toBe(expected);
        });

        it("sets annee_demande prioritary to date_debut", () => {
            mapMock.mockReturnValueOnce({ date_debut: new Date(2023, 10, 20), annee_demande: 2024 });
            const expected = 2024;
            const actual = DemarchesSimplifieesEntityAdapter.toSubvention(DEMANDE, MAPPING).annee_demande?.value;

            expect(actual).toBe(expected);
        });

        it("calls nestedToProviderValues", () => {
            // @ts-expect-error -- private spy
            const nestSpy = jest.spyOn(DemarchesSimplifieesEntityAdapter, "nestedToProviderValues");
            DemarchesSimplifieesEntityAdapter.toSubvention(DEMANDE, MAPPING);
            expect(nestSpy).toHaveBeenCalled();
        });
    });

    describe("toRawGrant", () => {
        it("should try to get joinKey", () => {
            DemarchesSimplifieesEntityAdapter.toRawGrant(DATA_ENTITIES[0], SCHEMAS[0]);
            expect(demarchesSimplifieesService.getJoinKey).toHaveBeenCalledWith({
                entity: DATA_ENTITIES[0],
                schema: SCHEMAS[0],
            });
        });

        it("should not define joinKey if not found", () => {
            jest.mocked(demarchesSimplifieesService.getJoinKey).mockReturnValueOnce(undefined);
            const expected = undefined;
            const actual = DemarchesSimplifieesEntityAdapter.toRawGrant(DATA_ENTITIES[0], SCHEMAS[0]).joinKey;
            expect(actual).toEqual(expected);
        });

        it("should return RawGrant", () => {
            const EJ = "EJ";
            jest.mocked(demarchesSimplifieesService.getJoinKey).mockReturnValueOnce(EJ);
            const expected = {
                provider: demarchesSimplifieesService.provider.id,
                type: "application",
                data: { entity: DATA_ENTITIES[0], schema: SCHEMAS[0] },
                joinKey: EJ,
            };
            const actual = DemarchesSimplifieesEntityAdapter.toRawGrant(DATA_ENTITIES[0], SCHEMAS[0]);
            expect(actual).toEqual(expected);
        });
    });

    describe("toCommon", () => {
        const ENTITY = { siret: "12345678901234" };
        const SCHEMA = {};

        it("adapts to proper format", () => {
            mapMock.mockReturnValueOnce({
                dateTransmitted: new Date("2022-01-07"),
                providerStatus: "sans_suite",
            });
            // @ts-expect-error mock
            const actual = DemarchesSimplifieesEntityAdapter.toCommon(ENTITY, SCHEMA);
            expect(actual).toMatchInlineSnapshot(`
                {
                  "exercice": 2022,
                  "siret": "12345678901234",
                  "statut": "InÃ©ligible",
                }
            `);
        });

        it("sets exercice thanks to dateTransmitted", () => {
            mapMock.mockReturnValueOnce({ dateTransmitted: new Date(2023, 10, 20) });
            const expected = 2023;
            const actual = DemarchesSimplifieesEntityAdapter.toCommon(DEMANDE, MAPPING).exercice;

            expect(actual).toBe(expected);
        });

        it("sets exercice prioritary to dateTransmitted", () => {
            mapMock.mockReturnValueOnce({ dateTransmitted: new Date(2023, 10, 20), exercice: 2024 });
            const expected = 2024;
            const actual = DemarchesSimplifieesEntityAdapter.toCommon(DEMANDE, MAPPING).exercice;

            expect(actual).toBe(expected);
        });

        it("sets siret from mapping if given", () => {
            const OTHER_SIRET = "09876543210987";
            mapMock.mockReturnValueOnce({ siret: OTHER_SIRET });
            const expected = OTHER_SIRET;
            const actual = DemarchesSimplifieesEntityAdapter.toCommon(DEMANDE, MAPPING).siret;

            expect(actual).toBe(expected);
        });

        it("sets siret from entity if nothing from mapping", () => {
            const expected = DEMANDE.siret;
            mapMock.mockReturnValueOnce({});
            const actual = DemarchesSimplifieesEntityAdapter.toCommon(DEMANDE, MAPPING).siret;

            expect(actual).toBe(expected);
        });
    });

    describe("mapSchema", () => {
        const ENTITY = { before: "a", siret: "SIRET" };
        const SCHEMA = {
            key: [
                { to: "after.nested", from: "before" },
                { to: "after.same", value: "toujoursPareil" },
            ],
        };
        const KEY = "key";

        it("adapts to proper format", () => {
            // @ts-expect-error mock
            const actual = DemarchesSimplifieesEntityAdapter.mapSchema(ENTITY, SCHEMA, KEY);
            expect(actual).toMatchInlineSnapshot(`
                {
                  "after": {
                    "nested": "a",
                    "same": "toujoursPareil",
                  },
                }
            `);
        });
    });

    describe("nestedToProviderValues", () => {
        const TO_PV = jest.fn();

        it("in an array call for each element", () => {
            const ARRAY = [0, 0, 0, 0];
            const expected = [1, 2, 3, 4];
            expected.forEach(v => TO_PV.mockReturnValueOnce(v));
            // @ts-expect-error -- test private method
            const actual = DemarchesSimplifieesEntityAdapter.nestedToProviderValues(ARRAY, TO_PV);
            expect(actual).toEqual(expected);
        });

        it("if neither object nor array return computed providerValue", () => {
            const VALUE = "toto";
            const expected = 1;
            TO_PV.mockReturnValueOnce(expected);
            // @ts-expect-error -- test private method
            const actual = DemarchesSimplifieesEntityAdapter.nestedToProviderValues(VALUE, TO_PV);
            expect(actual).toEqual(expected);
        });

        it("if object, call for each attribute", () => {
            const OBJECT = { a: 23, b: 34 };
            const expected = { a: 1, b: 2 };
            Object.values(expected).forEach(v => TO_PV.mockReturnValueOnce(v));
            // @ts-expect-error -- test private method
            const actual = DemarchesSimplifieesEntityAdapter.nestedToProviderValues(OBJECT, TO_PV);
            expect(actual).toEqual(expected);
        });
    });

    describe("toFlat", () => {
        const SCHEMA = "SCHEMA" as unknown as DemarchesSimplifieesSchema;
        const ADAPTED = {
            beneficiaryEstablishmentId: 12345678901234,
            provider: "producerSlug",
            applicationProviderId: "blablablaHash",
            budgetaryYear: 2024,
            ej: "567",
            depositDate: "2024-09-09",
            status: "statut Ã  adapter",
        };

        beforeAll(() => {
            mapMock.mockReturnValue(ADAPTED);
            mockAdaptStatus.mockReturnValue("statutAdaptÃ©");
        });

        afterAll(() => {
            mapMock.mockRestore();
            mockAdaptStatus.mockRestore();
        });

        it("maps schema", () => {
            DemarchesSimplifieesEntityAdapter.toFlat(DEMANDE, SCHEMA);
            expect(mapMock).toHaveBeenCalledWith(DEMANDE, SCHEMA, "flatSchema");
        });

        it("adapts status", () => {
            const expected = ADAPTED.status;
            DemarchesSimplifieesEntityAdapter.toFlat(DEMANDE, SCHEMA);
            // @ts-expect-error -- private attribute
            expect(providerAdapter.toStatusFactory).toHaveBeenCalledWith(
                DemarchesSimplifieesEntityAdapter._statusConversionArray,
            );
            expect(mockAdaptStatus).toHaveBeenCalledWith(expected);
        });

        it("adapts built attributes", () => {
            const ENTITY = { demarcheId: 987 } as DemarchesSimplifieesDataEntity;
            const actual = DemarchesSimplifieesEntityAdapter.toFlat(ENTITY, SCHEMA);
            expect(actual).toMatchSnapshot();
        });
    });
});
