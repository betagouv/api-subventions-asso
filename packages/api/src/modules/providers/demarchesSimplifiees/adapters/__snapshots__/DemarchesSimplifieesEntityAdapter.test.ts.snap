// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DemarchesSimplifieesEntityAdapter toFlat adapts built attributes 1`] = `
{
  "allocatorId": null,
  "allocatorIdType": null,
  "allocatorName": null,
  "applicationId": "demarches-simplifiees-987-blablablaHash",
  "applicationProviderId": "blablablaHash",
  "beneficiaryEstablishmentId": "12345678901234",
  "beneficiaryEstablishmentIdType": "siret",
  "budgetaryYear": 2024,
  "cofinancersIdType": null,
  "cofinancersNames": null,
  "cofinancingRequested": null,
  "confinancersId": null,
  "conventionDate": null,
  "decisionDate": null,
  "decisionReference": null,
  "depositDate": 2024-09-09T00:00:00.000Z,
  "ej": "567",
  "grantedAmount": null,
  "idRAE": null,
  "instructiveDepartementId": null,
  "instructiveDepartmentIdType": null,
  "instructiveDepartmentName": null,
  "joinKeyDesc": null,
  "joinKeyId": null,
  "managingAuthorityId": null,
  "managingAuthorityIdType": null,
  "managingAuthorityName": null,
  "nature": null,
  "object": null,
  "paymentCondition": null,
  "paymentConditionDesc": null,
  "paymentId": "12345678901234-567-2024",
  "paymentPeriodDates": null,
  "pluriannual": null,
  "pluriannualYears": null,
  "provider": "demarches-simplifiees-987",
  "requestYear": 2024,
  "requestedAmount": null,
  "scheme": null,
  "statusLabel": "statutAdapt√©",
  "subScheme": null,
  "subventionPercentage": null,
  "totalAmount": null,
  "ueNotification": null,
  "uniqueId": "demarches-simplifiees-987-blablablaHash-2024",
}
`;
